### PR TITLE
[aptos-rosetta] Add first account balance command to rosetta proxy

### PR DESCRIPTION
**Pull request by gregnazario**
## Motivation

Let's support [Rosetta](https://www.rosetta-api.org/docs/Reference.html).  I'm adopting and cleaning up the [proxy built for Diem](https://github.com/diem/rosetta-proxy).  Turns out a whole bunch of things needed to be cleaned up, so it wasn't easy to just port it directly without some changes.

This initial PR makes the webserver setup, and the simple "account balance" request.  More will come after this PR.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Tested locally against devnet.  The proxy works at getting the balance accordingly.

```
$ cargo run -p aptos-rosetta &
$ curl -X POST http://localhost:8080/account/balance/ -H "Content-Type: application/json" -d '{"network_identifier": {"blockchain":"aptos", "network": "devnet"}, "account_identifier": {"address":"57464DE88766306B2C6FCCD6835B617BC79FF189CCD1D2BD1C4D6781FB13211E"}}'

{"block_identifier":{"index":89596,"hash":"0x35a77fa23bff0b0ad9289c3f8b5e211a3de98440e2eb6c639926d3de6c1a0f88"},"balances":[{"value":"10000","currency":{"symbol":"APTOS","decimals":6}}]}
```

### CODE REVIEW

**Comment by jjleng** [comment]
So we decided to run Rosetta APIs as a standalone server?

**Comment by gregnazario** [comment]
> So we decided to run Rosetta APIs as a standalone server?

Yes, it is implemented as a proxy to a full node.

File: crates/aptos-rosetta/src/account.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    let txs = rest_client
+        .get_transactions(Some(state.version), Some(1))
+        .await
+        .map_err(|err| ApiError::AptosError(err.to_string()))?
+        .into_inner();
+    let tx = txs
+        .first()
+        .ok_or_else(|| ApiError::AptosError("Transaction not found".to_string()))?;
+
+    let txn_info = tx
```
Comment (davidiw) [inline crates/aptos-rosetta/src/account.rs:66]: tx or txn :P

File: crates/aptos-rosetta/src/account.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        .get_account_balance(address)
+        .await
+        .map_err(|err| ApiError::AptosError(err.to_string()))?;
+    let balance = response.into_inner();
+
+    // TODO: Cleanup to match reality
+    let balances = vec![Amount {
+        value: balance.coin.value.to_string(),
+        currency: Currency {
+            symbol: "APTOS".to_string(),
```
Comment (davidiw) [inline crates/aptos-rosetta/src/account.rs:85]: I thought we add this to on-chain info. https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/Coin.move#L63

**Review by davidiw — COMMENTED** [review]
coming along well

is it possible to point to the relevant documentation on what you're implementing (e.g., the URL of the current spec)

File: crates/aptos-rosetta/src/account.rs (hunk 1 above)
Comment (gregnazario) [reply crates/aptos-rosetta/src/account.rs:66]: Will fix them all to be txn.

File: crates/aptos-rosetta/src/account.rs (hunk 2 above)
Comment (gregnazario) [reply crates/aptos-rosetta/src/account.rs:85]: Didn't know about this.  Will fix.

File: crates/aptos-rosetta/src/account.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+/// Account balance command
+///
+/// [API Spec](https://www.rosetta-api.org/docs/AccountApi.html#accountbalance)
+async fn account_balance(
+    account_balance_request: AccountBalanceRequest,
+    server_context: RosettaContext,
+) -> Result<AccountBalanceResponse, ApiError> {
+    debug!("/account/balance");
```
[the remark this answers is not in the document]
Comment (gregnazario) [reply crates/aptos-rosetta/src/account.rs:39]: Can do, probably at trace level.

File: crates/aptos-rosetta/src/account.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    let network_identifier = account_balance_request.network_identifier;
+
+    check_network(network_identifier, &server_context)?;
+
+    // NOTE: we don't support lookups of account balance at specific blocks
+    if account_balance_request.block_identifier.is_some() {
+        return Err(ApiError::HistoricBalancesUnsupported);
+    }
```
[the remark this answers is not in the document]
Comment (gregnazario) [reply crates/aptos-rosetta/src/account.rs:48]: This is from the previous Diem implementation, I can update it accordingly (it's just currently not supported in this, not the underlying rest API)

File: crates/aptos-rosetta/src/account.rs (hunk 2 above)
Comment (gregnazario) [reply crates/aptos-rosetta/src/account.rs:85]: I'll fix this in a future PR, I need to make rest client updates for it.

**Comment by gregnazario** [comment]
> coming along well
> 
> is it possible to point to the relevant documentation on what you're implementing (e.g., the URL of the current spec)

It's in the code documentation: https://www.rosetta-api.org/docs/AccountApi.html

Coin operations are not needed, they're for blockchains like bitcoin.

File: crates/aptos-rosetta/src/account.rs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        server_context = ?server_context,
+        "account_balance for [{}]",
+        request.account_identifier.address
+    );
+
+    let network_identifier = request.network_identifier;
+
+    check_network(network_identifier, &server_context)?;
+
+    // TODO: support lookups of account balance at specific blocks for now
```
Comment (movekevin) [inline crates/aptos-rosetta/src/account.rs:49]: nit: "No" support ... for now?

File: crates/aptos-rosetta/src/account.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        .into_inner();
+    let txn = txns
+        .first()
+        .ok_or_else(|| ApiError::AptosError("Transaction not found".to_string()))?;
+
+    let txn_info = txn
+        .transaction_info()
+        .map_err(|err| ApiError::AptosError(err.to_string()))?;
+
+    let block_identifier = BlockIdentifier {
```
Comment (movekevin) [inline crates/aptos-rosetta/src/account.rs:70]: What are we returning this for? Is it for an indication that this is the balance at the current latest block number? If so, kinda nitpicky, but we should have a bit more documentation for what this endpoint is expected to return.

File: crates/aptos-rosetta/src/client.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        let request = AccountBalanceRequest {
+            network_identifier: chain_id.into(),
+            account_identifier: account.into(),
+            block_identifier: None,
+            currencies: None,
+        };
+        self.account_balance(&request).await
+    }
+
+    pub async fn account_balance(
```
Comment (movekevin) [inline crates/aptos-rosetta/src/client.rs:37]: What is the difference between this and account_balance_simple?

File: crates/aptos-rosetta/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+};
+
+mod account;
+pub mod client;
+pub mod common;
+pub mod error;
+pub mod types;
+
+pub const CURRENCY: &str = "APTOS";
+pub const NUM_DECIMALS: u64 = 6;
```
Comment (movekevin) [inline crates/aptos-rosetta/src/lib.rs:27]: Are these hardcoded just for tests?

File: crates/aptos-rosetta/src/account.rs (hunk 6 above)
Comment (gregnazario) [reply crates/aptos-rosetta/src/account.rs:70]: There are documentation links to the Rosetta spec on each API.  It's kinda debatable what each one has to do, so I'm extracting it from there.  I'm trying to not make too much duplicate documentation

File: crates/aptos-rosetta/src/client.rs (hunk 7 above)
Comment (gregnazario) [reply crates/aptos-rosetta/src/client.rs:37]: 'account_balance' is exactly what the spec requires for Rosetta.  'account_balance_simple' let's me skip the serialization and do it all in one place.

Think of it as a SDK on a spec, with an extension

File: crates/aptos-rosetta/src/lib.rs (hunk 8 above)
Comment (gregnazario) [reply crates/aptos-rosetta/src/lib.rs:27]: I'm going to come back and make these come from the actual rest API, but I'm trying to get the skeleton all in before I go ahead and start tweaking the APIs or writing custom transactions for my needs.

File: crates/aptos-rosetta/src/client.rs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+impl RosettaClient {
+    pub fn new(address: Url) -> RosettaClient {
+        RosettaClient {
+            address,
+            inner: ReqwestClient::new(),
+        }
+    }
+
+    pub async fn account_balance_simple(
```
Comment (davidiw) [inline crates/aptos-rosetta/src/client.rs:23]: might it be more reasonable to mark this as accessible via test only?

**Review by davidiw — APPROVED** [review]

File: crates/aptos-rosetta/src/client.rs (hunk 9 above)
Comment (gregnazario) [reply crates/aptos-rosetta/src/client.rs:23]: Yeah, we could do that, or I could just get rid of it, since it's causing confusion.

**Pull request closed without merging**